ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/TypeHintsAdornmentProvider.fs

```diff
@@ -0,0 +1,125 @@
+﻿namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Stages
+
+open System
+open System.Drawing
+open JetBrains.DocumentModel
+open JetBrains.ProjectModel
+open JetBrains.ReSharper.Daemon.Stages
+open JetBrains.ReSharper.Feature.Services.Daemon.Attributes
+open JetBrains.ReSharper.Plugins.FSharp
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.ReSharper.Feature.Services.Daemon
+open JetBrains.ReSharper.Plugins.FSharp.Daemon.Cs.Stages
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.Rider.Model
+open JetBrains.TextControl.DocumentMarkup
+open JetBrains.UI.RichText
+open FSharp.Compiler.SourceCodeServices
+open FSharp.Compiler.Layout
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree
+open JetBrains.ReSharper.Psi
+open JetBrains.ReSharper.Psi.Modules
+
+[<DaemonIntraTextAdornmentProvider(typeof<TypeHintsAdornmentProvider>)>]
+[<StaticSeverityHighlighting(Severity.INFO,
+     HighlightingGroupIds.IntraTextAdornmentsGroup,
+     AttributeId = AnalysisHighlightingAttributeIds.PARAMETER_NAME_HINT,
+     OverlapResolve = OverlapResolveKind.NONE,
+     ShowToolTipInStatusBar = false)>]
+type TypeHintHighlighting(text: RichText, range: DocumentRange) =
+    interface IHighlighting with
+        member x.ToolTip = null
+        member x.ErrorStripeToolTip = null
+        member x.IsValid() = not text.IsEmpty && not range.IsEmpty
+        member x.CalculateRange() = range
+
+    interface IHighlightingWithTestOutput with
+        member x.TestOutput = text.ToString()
+
+    member x.Text = text
+
+and [<SolutionComponent>] TypeHintsAdornmentProvider() =
+    interface IHighlighterIntraTextAdornmentProvider with
+        member x.CreateDataModel(highlighter) =
+            match highlighter.UserData with
+            | :? TypeHintHighlighting as thh ->
+                { new IIntraTextAdornmentDataModel with
```

Comment:
Could `TypeNameHintDataModel` type work here?

Reply:
If we used that then we would have to use `TypeNameHintHighlighting`, but we can't use that because we don't have an `IType` to navigate to :(

Reply:
Ah, it's pitty. We could easily convert `FSharpType` instances to `IType` ones if we had them on hand. Another reason to have a public API for it in FCS. :)

Reply:
If we use `CodeAnnotationHintHighlighting`, then because it sets `IsPreceding=true` on the data model, the hints end up in the wrong places when adding/removing blank lines between the `|>`

Reply:
I see, thanks for trying it out!